File: Pro_ODP_NET_for_Oracle_Database_11g-4541/c_sharp/Chapter 5/VArrays/ProductVArray.cs
using System;
using System.Data;
using System.Collections;
using Oracle.DataAccess.Client;
using Oracle.DataAccess.Types;

public class ProductVArray : IOracleCustomType, INullable
{
    //You will need to define a local array to hold the VARRAY elements. The data
    //type must correspond to the type declared in the VARRAY. You’ve defined a VARRAY
    //of VARCHAR2(10) values, hence your local array will hold String objects.
    [OracleArrayMapping()]
    public String[] Array;
    //The status array is used to store the status of an array index - whether the
    //element at the index is a NULL value or not.
    private OracleUdtStatus[] m_statusArray;

    public OracleUdtStatus[] StatusArray
    {
        get
        {
            return this.m_statusArray;
        }
        set
        {
            this.m_statusArray = value;
        }
    }
    
    private bool m_bIsNull;
    public bool IsNull
    {
        get
        {
            return m_bIsNull;
        }
    }
    
    public static ProductVArray Null
    {
        get
        {
            ProductVArray obj = new ProductVArray();
            obj.m_bIsNull = true;
            return obj;
        }
    }
    
    //The ToCustomObject method is required as part of the IOracleCustomType
    //implementation. It maps the retrieved VARRAY to the local array.
    public void ToCustomObject(OracleConnection con, IntPtr pUdt)
    {
        object objectStatusArray = null;
        Array = (String[])OracleUdt.GetValue(con, pUdt, 0, out objectStatusArray);
        m_statusArray = (OracleUdtStatus[])objectStatusArray;
    }
    
    //The FromCustomObject method is the opposite equivalent. It maps a local array to a
    //VARRAY
    public void FromCustomObject(OracleConnection con, IntPtr pUdt)
    {
        OracleUdt.SetValue(con, pUdt, 0, Array, m_statusArray);
    }
    
    public override string ToString()
    {
        if (m_bIsNull)
            return "ProductVArray.Null";
        else
        {
            string rtnstr = String.Empty;
            if (m_statusArray[0] == OracleUdtStatus.Null)
                rtnstr = "NULL";
            else
                rtnstr = Array.GetValue(0).ToString();
            for (int i = 1; i < m_statusArray.Length; i++)
            {
                if (m_statusArray[i] == OracleUdtStatus.Null)
                    rtnstr += "," + "NULL";
                else
                    rtnstr += "," + Array.GetValue(i).ToString();
            }
            return "ProductVArray(" + rtnstr + ")";
        }
    }
}

File: Pro_ODP_NET_for_Oracle_Database_11g-4541/vb_dot_net/Chapter5VBNET/VArrays/ProductVArray.vb
Imports System
Imports System.Data
Imports System.Collections
Imports Oracle.DataAccess.Client
Imports Oracle.DataAccess.Types

 _

Public Class ProductVArray
    Implements INullable

    <OracleArrayMapping()> _
    Public Array() As [String]
    'The status array is used to store the status of an array index - whether the
    'element at the index is a NULL value or not.
    Private m_statusArray() As OracleUdtStatus


    Public Property StatusArray() As OracleUdtStatus()
        Get
            Return Me.m_statusArray
        End Get
        Set(ByVal value As OracleUdtStatus())
            Me.m_statusArray = value
        End Set
    End Property

    Private m_bIsNull As Boolean

    Public ReadOnly Property IsNull() As Boolean Implements Oracle.DataAccess.Types.INullable.IsNull
        Get
            Return m_bIsNull
        End Get
    End Property


    Public Shared ReadOnly Property Null() As ProductVArray
        Get
            Dim obj As New ProductVArray()
            obj.m_bIsNull = True
            Return obj
        End Get
    End Property


    'The ToCustomObject method is required as part of the IOracleCustomType
    'implementation. It maps the retrieved VARRAY to the local array.
    Public Sub ToCustomObject(ByVal con As OracleConnection, ByVal pUdt As IntPtr)
        Dim objectStatusArray As Object = Nothing
        Array = CType(OracleUdt.GetValue(con, pUdt, 0, objectStatusArray), [String]())
        m_statusArray = CType(objectStatusArray, OracleUdtStatus())
    End Sub 'ToCustomObject


    'The FromCustomObject method is the opposite equivalent. It maps a local array to a
    'VARRAY
    Public Sub FromCustomObject(ByVal con As OracleConnection, ByVal pUdt As IntPtr)
        OracleUdt.SetValue(con, pUdt, 0, Array, m_statusArray)
    End Sub 'FromCustomObject


    Public Overrides Function ToString() As String
        If m_bIsNull Then
            Return "ProductVArray.Null"
        Else
            Dim rtnstr As String = [String].Empty
            If m_statusArray(0) = OracleUdtStatus.Null Then
                rtnstr = "NULL"
            Else
                rtnstr = Array.GetValue(0).ToString()
            End If
            Dim i As Integer
            For i = 1 To m_statusArray.Length - 1
                If m_statusArray(i) = OracleUdtStatus.Null Then
                    rtnstr += "," + "NULL"
                Else
                    rtnstr += "," + Array.GetValue(i).ToString()
                End If
            Next i
            Return "ProductVArray(" + rtnstr + ")"
        End If
    End Function 'ToString


End Class 'ProductVArray